### PR TITLE
[InstCombine][FP] Fix nnan preservation for transform fcmp + sel => fmax/fmin

### PR DESCRIPTION
**Pull request by rajatbajpai**
Preserve `nnan` constraint only if present on both `fcmp` and `select`.

Alive2: https://alive2.llvm.org/ce/z/ZNDjzt

### CODE REVIEW

**Comment by rajatbajpai** [comment]
@arsenm Could you please review this fix?
Original discussion about fixing this issue: https://github.com/llvm/llvm-project/pull/106492#discussion_r1781035045

**Review by nikic — COMMENTED** [review]
It doesn't look like the nnan is required, you just can't always preserve it? https://alive2.llvm.org/ce/z/4Dj45b

**Comment by rajatbajpai** [comment]
Right now we try to preserve it even if `fcmp` doesn't have that flag. https://godbolt.org/z/5vPjrWT95

**Review by dtcxzyw — CHANGES_REQUESTED** [review]
https://alive2.llvm.org/ce/z/4PKqnZ

You should drop `nnan` in the new `maxnum/minnum` instead of blocking this transform.

**Comment by dtcxzyw** [comment]
> https://alive2.llvm.org/ce/z/4PKqnZ
> 
> You should drop `nnan` in the new `maxnum/minnum` instead of blocking this transform.

Ignore me. @nikic already pointed out this.

File: llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp [hunk 1]
```diff
@@ -3864,17 +3864,25 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
...
     // minnum/maxnum intrinsics.
     if (SIFPOp->hasNoNaNs() && SIFPOp->hasNoSignedZeros()) {
       Value *X, *Y;
-      if (match(&SI, m_OrdOrUnordFMax(m_Value(X), m_Value(Y))))
-        return replaceInstUsesWith(
-            SI, Builder.CreateBinaryIntrinsic(Intrinsic::maxnum, X, Y, &SI));
+      if (match(&SI, m_OrdOrUnordFMax(m_Value(X), m_Value(Y)))) {
+        Value *BinIntr =
+            Builder.CreateBinaryIntrinsic(Intrinsic::maxnum, X, Y, &SI);
+        cast<Instruction>(BinIntr)->setHasNoNaNs(FCmp->hasNoNaNs());
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:3876]: `BinIntr` may be folded into a constant by IRBuilder.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Comment by rajatbajpai** [comment]
Please merge the PR if it looks okay. Thanks!

**Comment by nikic** [comment]
Can you please update the PR description / proof to match the new implementation?

**Comment by rajatbajpai** [comment]
> Can you please update the PR description / proof to match the new implementation?

Done, thanks!

**Pull request merged**